### PR TITLE
feat: support inserting local image

### PR DESCRIPTION
**Pull request by LucasXu0**
### Feature Preview
Screenshot 2023-06-30 at 14 30 10

* support inserting local image.
* add image menu
  * align
  * copy link
  * delete

---

#### PR Checklist

- [x] My code adheres to the [AppFlowy Style Guide](https://appflowy.gitbook.io/docs/essential-documentation/contribute-to-appflowy/software-contributions/submitting-code/style-guides)
- [x] I've listed at least one issue that this PR fixes in the description above.
- [ ] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [x] All existing tests are passing.

### CODE REVIEW

File: frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/image/image_menu.dart [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+            onTap: copyImageLink,
+          ),
+          const HSpace(4),
+          _ImageAlignButton(
+            node: widget.node,
+            state: widget.state,
+          ),
+          const _Divider(),
+          _ImageDeleteButton(
+            onTap: () => deleteImage(),
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/image/image_menu.dart:56]:
```suggestion
            onTap: deleteImage,
```

File: frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/image/image_menu.dart [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+  }
+
+  void copyImageLink() {
+    final url = widget.node.attributes[ImageBlockKeys.url];
+    if (url != null) {
+      Clipboard.setData(ClipboardData(text: url));
+      ScaffoldMessenger.of(context).showSnackBar(
+        SnackBar(
+          content: FlowyText(
+            LocaleKeys.document_plugins_image_copyToPasteBoard.tr(),
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/image/image_menu.dart:71]:
```suggestion
            LocaleKeys.document_plugins_image_copiedToClipBoard.tr(),
```

File: frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/image/image_menu.dart [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+  State<_ImageAlignButton> createState() => _ImageAlignButtonState();
+}
+
+const interceptorKey = 'image-align';
+
+class _ImageAlignButtonState extends State<_ImageAlignButton> {
+  late final gestureInterceptor = SelectionGestureInterceptor(
+    key: interceptorKey,
+    canTap: (details) => false,
+  );
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/image/image_menu.dart:126]: Does this have to be `late`?

File: frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/image/image_menu.dart [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+  String get align => widget.node.attributes['align'] ?? 'center';
+  final popoverController = PopoverController();
+  late final EditorState editorState;
+
+  @override
+  void initState() {
+    super.initState();
+
+    editorState = context.read<EditorState>();
+    // preventMenuClose();
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/image/image_menu.dart:137]: Remove

File: frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/image/image_menu.dart [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return IgnoreParentGestureWidget(
+      child: AppFlowyPopover(
+        onClose: () {
+          allowMenuClose();
+        },
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/image/image_menu.dart:153]:
```suggestion
        onClose: allowMenuClose,
```

File: frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/image/image_selection_menu.dart [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    final container = Overlay.of(context);
+    showImageMenu(
+      container,
+      editorState,
+      menuService,
+      onInsertImage: (url) async {
+        // if the url is http, we can insert it directly
+        // otherwise, if it's a file url, we need to copy the file to the app's document directory
+
+        final regex = RegExp('^(http|https)://');
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/image/image_selection_menu.dart:29]: Should we move useful and reusable RegExp patterns into `flowy_infra`? To not have multiple of the same purpose RegExp patterns that might contain different rules and expect the same result. (Consistent behvaior)

File: frontend/appflowy_flutter/lib/plugins/document/presentation/more/cubit/document_appearance_cubit.dart [hunk 7]
```diff
@@ -25,6 +25,11 @@ class DocumentAppearanceCubit extends Cubit<DocumentAppearance> {
   void fetch() async {
     final prefs = await SharedPreferences.getInstance();
     final fontSize = prefs.getDouble(_kDocumentAppearanceFontSize) ?? 16.0;
+
+    if (isClosed) {
+      return;
+    }
+
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/more/cubit/document_appearance_cubit.dart:32]:
```suggestion
    if (isClosed) {
      return;
    }

    final prefs = await SharedPreferences.getInstance();
    final fontSize = prefs.getDouble(_kDocumentAppearanceFontSize) ?? 16.0;
```

File: frontend/appflowy_flutter/lib/plugins/document/presentation/more/cubit/document_appearance_cubit.dart [hunk 8]
```diff
@@ -35,6 +40,11 @@ class DocumentAppearanceCubit extends Cubit<DocumentAppearance> {
   void syncFontSize(double fontSize) async {
     final prefs = await SharedPreferences.getInstance();
     prefs.setDouble(_kDocumentAppearanceFontSize, fontSize);
+
+    if (isClosed) {
+      return;
+    }
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/more/cubit/document_appearance_cubit.dart:46]:
```suggestion
    if (isClosed) {
      return;
    }

    final prefs = await SharedPreferences.getInstance();
    prefs.setDouble(_kDocumentAppearanceFontSize, fontSize);
```

**Review by Xazin — APPROVED** [review]
LGTM - Minor comments

File: frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/image/image_selection_menu.dart (hunk 6 above)
Comment (LucasXu0) [reply frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/image/image_selection_menu.dart:29]: Fair enough. I agree that we should move all these `RegExp` into `flowy_infra`. I will do that in the next PR to refactor all of them.

**Pull request merged**